system_zero/src/alu/rotate_shift.rs

```diff
@@ -0,0 +1,596 @@
+    values[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_0] = F::from_canonical_u32(shifted_hi_0);
+    values[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_1] = F::from_canonical_u32(shifted_hi_1);
+
+    if shifted_lo_1 != u32::MAX {
+        let diff = F::from_canonical_u32(u32::MAX - shifted_lo_1);
+        let inv = diff.inverse();
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_AUX_0] = inv;
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_AUX_1] = diff * inv;
+    } else {
+        // shifted_lo_0 must be zero, so this is unused.
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_AUX_0] = F::ZERO;
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_AUX_1] = F::ZERO;
+    }
+    if shifted_hi_1 != u32::MAX {
+        let diff = F::from_canonical_u32(u32::MAX - shifted_hi_1);
+        let inv = diff.inverse();
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_AUX_0] = inv;
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_AUX_1] = diff * inv;
+    } else {
+        // shifted_hi_0 must be zero, so this is unused.
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_AUX_0] = F::ZERO;
+        values[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_AUX_1] = F::ZERO;
+    }
+
+    // the input and output as u64s
+    let input = (input_hi << 32) | input_lo;
+    let delta = delta as u32;
+    let output = match op {
+        IS_ROTATE_LEFT => input.rotate_left(delta),
+        IS_ROTATE_RIGHT => input.rotate_right(delta),
+        IS_SHIFT_LEFT => input << delta,
+        IS_SHIFT_RIGHT => input >> delta,
+        IS_ARITH_SHIFT_RIGHT => (input as i64 >> delta) as u64,
+        _ => panic!("unrecognized rotate/shift instruction code"),
+    };
+
+    // Output in base 2^16.
+    values[COL_ROTATE_SHIFT_OUTPUT_0] = F::from_canonical_u32(output as u32);
+    values[COL_ROTATE_SHIFT_OUTPUT_1] = F::from_canonical_u32((output >> 32) as u32);
+}
+
+fn eval_rotate_shift<F: Field, P: PackedField<Scalar = F>>(
+    lv: &[P; NUM_COLUMNS],
+    yield_constr: &mut ConstraintConsumer<P>,
+    filter: P,
+) -> (P, P, P, P, P, P, P) {
+    // Input
+    let input_lo = lv[COL_ROTATE_SHIFT_INPUT_LO];
+    let input_hi = lv[COL_ROTATE_SHIFT_INPUT_HI];
+
+    // Delta is the shift/rotate displacement; exp is delta mod 32 or
+    // (32 - delta mod 32) mod 32, depending on whether the operation
+    // direction is left or right.
+    let exp_bits = COL_ROTATE_SHIFT_EXP_BITS.map(|r| lv[r]);
+    let top_bit = lv[COL_ROTATE_SHIFT_DELTA_DIV32];
+
+    let pow_exp_aux_0 = lv[COL_ROTATE_SHIFT_POW_EXP_AUX_0];
+    let pow_exp_aux_1 = lv[COL_ROTATE_SHIFT_POW_EXP_AUX_1];
+    let pow_exp_aux_2 = lv[COL_ROTATE_SHIFT_POW_EXP_AUX_2];
+    let pow_exp = lv[COL_ROTATE_SHIFT_POW_EXP];
+
+    let shifted_lo_0 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_0];
+    let shifted_lo_1 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_LO_1];
+    let shifted_hi_0 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_0];
+    let shifted_hi_1 = lv[COL_ROTATE_SHIFT_DISPLACED_INPUT_HI_1];
```

Comment:
When I read `shifted_lo`, I expect the `lo` part of the `shifted` value. But this is the `lo` part that has been `shifted`. I think `lo_shifted` would be less confusing, at least for me.